Issue 3582 configure Netty shutdown quiet period and shutdown timeout

wetted opened the pull request:
Fixes #3582 - adding feature to configure the shutdown quiet period and shutdown timeout for Netty event loop groups and http server

wetted committed:
Issue 3582 - adding feature to configure the shutdown quiet period and shutdown timeout for Netty event loop groups and http server

wetted committed:
Issue 3582 - adding feature to configure the shutdown quiet period and shutdown timeout for Netty event loop groups and http server

wetted committed:
Issue 3582
- adding feature to configure the shutdown quiet period and shutdown timeouts
- adding overloaded versions of HttpClient create; RxStreamingHttpClient create; and RxHttpClientFactory createClient and createStreamingClient to accept a HttpClientConfiguration instance, for cases where the client is created outside the Micronaut application context

wetted committed:
Issue 3582
- adding feature to configure the shutdown quiet period and shutdown timeouts
- adding overloaded versions of HttpClient create; RxStreamingHttpClient create; and RxHttpClientFactory createClient and createStreamingClient to accept a HttpClientConfiguration instance, for cases where the client is created outside the Micronaut application context

wetted committed:
Issue 3582
- revise section 6.27.1 of the documentaion to add clarification about Netty server configuration
- add deprecation warning in NettyHttpServer about preferring Netty event loop groups configurations

jameskleeh commented on http-netty/src/main/java/io/micronaut/http/netty/channel/DefaultEventLoopGroupRegistry.java:
```diff
@@ -56,29 +66,49 @@ public DefaultEventLoopGroupRegistry(EventLoopGroupFactory eventLoopGroupFactory
+                long timeout = configuration.getShutdownTimeout().toMillis();
+                eventLoopGroup.shutdownGracefully(quietPeriod, timeout, TimeUnit.MILLISECONDS);
+            } catch (Throwable t) {
+                LOGGER.warn("Error shutting down EventLoopGroup: {}", t.getMessage(), t);
```
This should be surrounded with `if (LOGGER.isWarnEnabled())`

jameskleeh commented on http-netty/src/main/java/io/micronaut/http/netty/channel/DefaultEventLoopGroupRegistry.java:
```diff
@@ -42,9 +48,13 @@
 @Internal
 @BootstrapContextCompatible
 public class DefaultEventLoopGroupRegistry implements EventLoopGroupRegistry {
+    private static final Logger LOGGER = LoggerFactory.getLogger(DefaultEventLoopGroupRegistry.class);
```
`LOG` here is generally preferred over `LOGGER`

jameskleeh commented on http-client-core/src/main/java/io/micronaut/http/client/RxStreamingHttpClient.java:
```diff
@@ -59,4 +59,16 @@
+     * @param configuration The client configuration
+     * @return The client
+     */
+    static RxStreamingHttpClient create(URL url, HttpClientConfiguration configuration) {
```
Please add @since 2.2.0 here

jameskleeh commented on http-client-core/src/main/java/io/micronaut/http/client/RxHttpClientFactory.java:
```diff
@@ -43,4 +53,14 @@
+     * @param configuration The client configuration
+     * @return The client
+     */
+    RxStreamingHttpClient createStreamingClient(@Nullable URL url, HttpClientConfiguration configuration);
```
Please add @since 2.2.0 here

jameskleeh commented on http-client-core/src/main/java/io/micronaut/http/client/RxHttpClientFactory.java:
```diff
@@ -35,6 +35,16 @@
+     * @param configuration the client configuration
+     * @return The client
+     */
+    RxHttpClient createClient(@Nullable URL url, HttpClientConfiguration configuration);
```
Please add @since 2.2.0 here

jameskleeh commented on http-client-core/src/main/java/io/micronaut/http/client/HttpClientConfiguration.java:
```diff
@@ -609,6 +649,26 @@ static RxHttpClient createClient(@Nullable URL url) {
+     * @return The client
+     */
+    @Internal
+    static RxStreamingHttpClient createStreamingClient(@NonNull URL url, HttpClientConfiguration configuration) {
```
Please add @since 2.2.0 here

jameskleeh commented on http-client-core/src/main/java/io/micronaut/http/client/HttpClientConfiguration.java:
```diff
@@ -586,19 +619,26 @@ public Proxy resolveProxy(boolean isSsl, String host, int port) {
+     * @return The client
+     */
+    @Internal
+    static RxHttpClient createClient(@Nullable URL url, HttpClientConfiguration configuration) {
```
Please add @since 2.2.0 here

jameskleeh commented on http-client-core/src/main/java/io/micronaut/http/client/HttpClient.java:
```diff
@@ -233,4 +233,16 @@ default HttpClient refresh() {
+     * @param configuration the client configuration
+     * @return The client
+     */
+    static HttpClient create(@Nullable URL url, HttpClientConfiguration configuration) {
```
Please add @since 2.2.0 here

jameskleeh committed:
Update threadPools.adoc

wetted committed:
Issue 3582
- adding feature to configure the shutdown quiet period and shutdown timeouts
- adding overloaded versions of HttpClient create; RxStreamingHttpClient create; and RxHttpClientFactory createClient and createStreamingClient to accept a HttpClientConfiguration instance, for cases where the client is created outside the Micronaut application context
